Support automatic attachment of export devices

rmol opened the pull request:
Adds udev rules and scripts to `sys-usb`. Modifies the qubes.USB RPC policy and adds qubes.USBAttach.

## Testing

- Check out this branch in your development Qube.
- In your cloned repo in `dom0`:
  - `make clone`
  - `make clean && make all`
  - Check that `/etc/qubes-rpc/policy/qubes.USB` contains:
    ```
    ### BEGIN securedrop-workstation ###
    sd-export-usb sys-usb allow
    @anyvm @tag:sd-workstation deny
    ### END securedrop-workstation ###
    @anyvm @anyvm ask
    ```
  - Check that `/etc/qubes-rpc/policy/qubes.USBAttach` contains:
    ```
    ### BEGIN securedrop-workstation ###
    sys-usb sd-export-usb allow,user=root
    @anyvm @tag:sd-workstation deny
    ### END securedrop-workstation ###
    @anyvm @anyvm ask
    ```
- In `sys-usb`, confirm the existence of:
  - `/etc/udev/rules.d/99-sd-export-usb.rules`
  - `/usr/local/bin/sd-attach-export-disk`
  - `/usr/local/bin/sd-attach-export-printer`

Plug in a valid LUKS export stick. It should be automatically attached to `sd-export-usb`, and an export from the client or manually via `qvm-open-in-vm sd-export-usb "filename"` should succeed.

creviera commented:
> Plug in a valid LUKS export stick. It should be automatically attached to sd-export-usb, and an export from the client or manually via qvm-open-in-vm sd-export-usb "filename" should succeed.

it also looks like this attaches drives that are not encrypted right? just want to double check

rmol commented:
That's right; it attaches any storage or printer device. The script invoked by the udev rule could check the partition table and label, but they're supposed to be quick. Also, I'm not sure how we'd surface any errors in a nice way. I think it's better to do any validation in `sd-export-usb`.

emkll commented on dom0/sys-usb.top:
```diff
@@ -0,0 +1,6 @@
+# vim: set syntax=yaml ts=2 sw=2 sts=2 et :
+
+base:
+  dom0:
```
I think it shouldn't be dom0 here, but rather sys-usb. I think this will resolve the issue i've experienced where it copies the files to dom0

rmol replied:
Yep. Major testing fail when I tried to get the `sys-usb` added to the entire Salt run. Fixed.

emkll commented on dom0/sd-dom0-qvm-rpc.sls:
```diff
@@ -76,7 +76,23 @@ dom0-rpc-qubes.USB:
     - marker_end: "### END securedrop-workstation ###"
     - content: |
-        $anyvm $tag:sd-workstation deny
+        sd-export-usb sys-usb allow
```
these policy changes result in failing RPC tests. Could you please update the test policies in https://github.com/freedomofpress/securedrop-workstation/blob/307-udev-export-devices/tests/vars/qubes-rpc.yml ?

rmol replied:
Updated.

emkll commented on dom0/sys-usb.sls:
```diff
@@ -0,0 +1,31 @@
+
+sd-udev-rules:
+  file.managed:
+    - name: /etc/udev/rules.d/99-sd-export-usb.rules
```
These rules will not persist accross reboots, since sys-usb is an AppVM. I just did a quick investigation and it appears `/usr/lib/udev/rules.d/` also does not persist reboots. We should investigate how qubes persists their custom udev rules, or perhaps add a script in /rw/ to ensure persistence.

rmol replied:
Now copying the rules under `/rw/config` and installing them via `/rw/config/rc.local`.

emkll commented on dom0/sd-dom0-qvm-rpc.sls:
```diff
@@ -76,7 +76,23 @@ dom0-rpc-qubes.USB:
-        $anyvm $tag:sd-workstation deny
+        sd-export-usb sys-usb allow
+        @anyvm @tag:sd-workstation deny
+dom0-rpc-qubes.ensure.USBAttach:
```
I am not sure what this step does, does this ensure the default USBAttach policy is present, and otherwise populates it with the default specified here? If that is the case, we may want to require this step prior to running the `dom0-rpc-qubes.USBAttach` step to prevent that change from being applied *before* this check is performed.

rmol replied:
Added the `require` to ensure correct ordering.

rmol commented:
Corrected the target in the `sys-usb` Salt config. Added tests to catch my slop; sorry for wasting your time. 

Parroted the [Qubes YubiKey configuration](https://www.qubes-os.org/doc/yubi-key/) you found to get the udev rules to survive reboots.

I'm not sure yet how best to undo this config. In addition to just removing it from `sys-usb`, it would need to be cleaned out of `/srv/salt` in dom0 -- I ran into this after switching to another issue; I had to manually excise these changes. We almost need Salt migrations. :slightly_smiling_face: I don't like the idea of littering the Salt config with one-time removal of obsolete files there, but don't know of a better way yet.

emkll commented:
> I'm not sure yet how best to undo this config. In addition to just removing it from sys-usb, it would need to be cleaned out of /srv/salt in dom0

good point, hadn't thought of that. We may need to handle this through conditional logic, maybe in an env var or config.json?.

conorsch commented:
> We almost need Salt migrations. slightly_smiling_face I don't like the idea of littering the Salt config with one-time removal of obsolete files there, but don't know of a better way yet.

Also relevant:

* https://github.com/freedomofpress/securedrop-workstation/pull/323#pullrequestreview-316576701
* https://github.com/freedomofpress/securedrop-workstation/pull/331#pullrequestreview-316342920

So it sounds like we _do_ need a more robust "clean" action. For instance, we leave the `sys-firewall` customizations (for the dom0 RPM repo) in place after `make clean` is run. Will open another issue.

kushaldas commented:
In my mind this is still a lot of code + configuration to maintain for auto attach. Doing it via `qubesadmin` API (Python code) seems to be much more manageable for future.

Here we are adding a complete new qrexec service and we will have to manage the config etc.

conorsch commented:
>  Will open another issue.

See #337.

emkll commented:
There's also an issue with the Qubes USB Widget, that was raised in the November 14th engineering meeting: 
- When a device is auto-attached, it is not registered as attached by the USB GUI widget (bolded and shown as attached to `sd-export-usb`)
- Using this Qubes USB widget to attach a device to a given VM will result in an error:
```
ERROR
Attaching device <device> to <vm> Error:DeviceAlreadyAttached: "Device sys-usb:2-1 already attached to sd-export-usb"
```
Manually running `qvm-usb detach sys-usb:2-1` resolves the issue (and a user may now attach that device to whichever Qube they choose), but requires a manual step.

rmol commented:
@emkll  Thanks. I'll need to investigate the boot-time udev handling. The problem with the devices widget seems to be that it isn't receiving or responding to the `device-list-changed:usb` events fired by `qubesusbproxy`. I've confirmed that the changes to the database under `/qubes-usb-devices` are being noticed when the device is attached, so I think the problem lies only in the widget.

emkll commented:
Another side effect of the changes introduced here that I forgot to note, is the increased likelihood of having multiple relevant (mass storage or printer) devices connected. We already do have an issue to track this: https://github.com/freedomofpress/securedrop-export/issues/25

redshiftzero commented on Makefile:
```diff
@@ -20,6 +20,10 @@ qubes-rpc: prep-salt ## Places default deny qubes-rpc policies for sd-svs and sd
 	sudo qubesctl top.enable sd-dom0-qvm-rpc
 	sudo qubesctl --show-output --targets sd-dom0-qvm-rpc state.highstate
 
+sys-usb: prep-salt ## Adds udev rules and scripts to sys-usb
```
I think we'll also want to add `sys-usb` to `scripts/list-vm` such that `make all` runs this target. I (otherwise successfully!) provisioned on this but noticed the customizations were missing in `sys-usb` until I ran `make sys-usb` separately

rmol replied:
I added `sys-usb` to `scripts/provision-all`.

redshiftzero commented:
I tested this today and the auto-attachment logic is working nicely for me. I can reproduce that the devices widget currently used in qubes does not reflect the auto-attachment of USBs to our sd-export-usb VM, but regarding the last issue, @rmol has some wip resolving the issue (this branch currently has some experimental UI updates) in the [qubes devices widget](https://github.com/QubesOS/qubes-desktop-linux-manager/compare/master...rmol:change-device-widget-update-strategy) which resolves for me (next step is he's going to make a PR upstream). 

If anyone wants to test or make changes to the devices widget here's instructions on how to get started: 

1. git clone your fork https://github.com/QubesOS/qubes-desktop-linux-manager in your sd-dev AppVM and check out the relevant branch
2. in dom0 stop the existing devices widget: `systemctl --user stop qubes-widget@qui-devices`
3. in dom0 assuming you have cloned the repo in your homedir in sd-dev: `qvm-run --pass-io sd-dev 'tar -c -C /home/user qubes-desktop-linux-manager' | tar xvf -`
4. in dom0 run `python3 -m qui.tray.devices`

rmol commented:
PR for the devices widget: https://github.com/QubesOS/qubes-desktop-linux-manager/pull/87

emkll commented:
Due to changes in https://github.com/freedomofpress/securedrop-workstation/pull/387, this will need to be rebased on latest master. We should also deny-by-default outbound USB policies ( `@tag:sd-workstation @anyvm deny`) as part of this PR as well.

Upstream fix for the device widget will be released in qubes-usb-proxy 1.0.23 which is still currently in testing repos.

rmol commented:
Rebased, outbound denial added to policies.

emkll commented:
Apologies, @rmol this will need to once again be rebased on latest master: due to changes introduced #407, sd-export-usb was renamed to sd-devices.

rmol commented:
I rebased, updated the naming, it seems to be working fine and [in about a week the Qubes devices widget should be correctly showing the state of automatically attached devices](https://github.com/QubesOS/qubes-issues/issues/5505#issuecomment-577406796).

rmol commented:
The upstream fix has [made it to stable](https://github.com/QubesOS/qubes-issues/issues/5505#issuecomment-578966425).

emkll commented on rpm-build/SPECS/securedrop-workstation-dom0-config.spec:
```diff
@@ -56,6 +57,7 @@ install -m 644 sd-proxy/* %{buildroot}/srv/salt/sd/sd-proxy/
 install -m 644 sd-whonix/* %{buildroot}/srv/salt/sd/sd-whonix/
 install -m 644 sd-workstation/* %{buildroot}/srv/salt/sd/sd-workstation/
 install -m 644 sys-firewall/* %{buildroot}/srv/salt/sd/sys-firewall/
+install sys-usb/* %{buildroot}/srv/salt/sd/sys-usb/
```
Could we explicitly set the mask here for consistency?

rmol replied:
Will do.

emkll commented on sys-usb/99-sd-devices.rules:
```diff
@@ -0,0 +1,6 @@
+# Class 08 == storage, subclass 06 == SCSI
```
thanks for these very helpful comments

emkll commented on dom0/sd-dom0-qvm-rpc.sls:
```diff
@@ -83,8 +83,27 @@ dom0-rpc-qubes.USB:
+  file.managed:
+    - name: /etc/qubes-rpc/policy/qubes.USBAttach
+    - contents: |
+        @anyvm @anyvm ask
```
If I understand correctly, this is to ensure the qubes.USBAttach policy is populated with default Qubes content, and the next step will ensure we apply rpc policies to SD-specific VMs?

rmol replied:
Exactly.

rmol commented:
Good catches and suggestions. I: 

- renamed the `sys-usb` make target to `add-usb-autoattach`
- added a `remove-usb-autoattach` make target
- included the removal logic in `sd-clean-all.sls`, so now `make clean` cleans up `sys-usb`
- explicitly set permissions of autoattach files in RPM specfile

emkll commented on dom0/sd-usb-autoattach-remove.sls:
```diff
@@ -0,0 +1,9 @@
+  cmd.run:
+    - names:
+      - qvm-run sys-usb 'sudo rm -f /etc/udev/rules.d/99-sd-devices.rules'
+      - qvm-run sys-usb 'sudo rm -f /rw/config/sd/etc/udev/rules.d/99-sd-devices.rules'
```
nit: why not `rm -rf  /rw/config/sd` here?

rmol replied:
Just being conservative and only deleting what was added for auto attachment. I can't think of what else we might add to the USB Qube, so it'd probably be fine to remove the whole `sd` directory, but this should be safer. Did I miss something? Is the directory structure being left behind a problem?

emkll replied:
Fair point on it being safer to delete explicit files, if we do add more files in the future we can re-evaluate.

emkll commented on Makefile:
```diff
@@ -19,6 +19,12 @@ clone: assert-dom0 ## Pulls the latest repo from work VM to dom0
 	sudo qubesctl --show-output --targets sd-dom0-qvm-rpc state.highstate
 
+add-usb-autoattach: prep-salt ## Adds udev rules and scripts to sys-usb
+	sudo qubesctl --show-output --skip-dom0 --targets sys-usb state.highstate
```
nit: On workstations built prior to this being merged, users will need to `make prep-dom0`, to copy the RPC policies in place. This is a one-time thing and not a huge deal, but calling prep-dom0 target before this here should resolve

rmol replied:
Will add that.
